lib/Macros/Sources/SwiftMacros/DebugDescriptionMacro.swift

```diff
@@ -0,0 +1,461 @@
+        // `debugDescription` takes priority: skip `description` if `debugDescription` also exists.
+        if propertyName == "description" && properties["debugDescription"] != nil {
+            return []
+        }
+
+        guard let moduleName = context.moduleName(of: declaration) else {
+            assertionFailure("could not determine module name from fileID")
+            return []
+        }
+
+        var typeIdentifier: String
+        if let typeParameters = declaration.asProtocol(WithGenericParametersSyntax.self)?.genericParameterClause?.parameters, typeParameters.count > 0 {
+            let typePatterns = Array(repeating: ".+", count: typeParameters.count).joined(separator: ",")
+            // A regex matching that matches the generic type.
+            typeIdentifier = "^\(moduleName)\\.\(typeName)<\(typePatterns)>"
+        } else if declaration.is(ExtensionDeclSyntax.self) {
+            // When attached to an extension, the type may or may not be a generic type.
+            // This regular expression handles both cases.
+            typeIdentifier = "^\(moduleName)\\.\(typeName)(<.+>)?$"
+        } else {
+            typeIdentifier = "\(moduleName).\(typeName)"
+        }
+
+        let computedProperties = properties.values.filter(\.isComputedProperty).compactMap(\.name)
+
+        return ["@_DebugDescriptionProperty(\"\(raw: typeIdentifier)\", \(raw: computedProperties))"]
+    }
+}
+
+/// An internal macro which performs which converts compatible description implementations to an LLDB type
+/// summary.
+///
+/// The LLDB type summary record is emitted into a custom section, which LLDB loads from at debug time.
+///
+/// Conversion has limitations, primarily that expression evaluation is not supported. If a description
+/// property calls another function, it cannot be converted. When conversion cannot be performed, an error
+/// diagnostic is emitted.
+///
+/// Note: There is one ambiguous case: computed properties. The macro can identify some, but not all, uses of
+/// computed properties. When a computed property cannot be identified at compile time, LLDB will emit a
+/// warning at debug time.
+///
+/// See https://lldb.llvm.org/use/variable.html#type-summary
+extension _DebugDescriptionPropertyMacro: PeerMacro {
+    public static func expansion<Decl, Context>(
+        of node: AttributeSyntax,
+        providingPeersOf declaration: Decl,
+        in context: Context
+    )
+    throws -> [DeclSyntax]
+    where Decl: DeclSyntaxProtocol, Context: MacroExpansionContext
+    {
+        guard case .argumentList(let argumentList) = node.arguments else {
+            assertionFailure("expected argumentList")
+            return []
+        }
+
+        let arguments = argumentList.map(\.expression)
+        guard let typeIdentifier = String(expr: arguments[0]),
+              let computedProperties = Array<String>(expr: arguments[1]) else {
+            assertionFailure("_DebugDescriptionProperty called incorrectly")
```

Comment:
In general, I would prefer to emit errors here instead of hitting `assertionFailure`s. As I mentioned above, somebody might redeclare `_DebugDescriptionPropertyMacro`.

Reply:
I'm not sure what scenario would result in broken internals being used for real. Also, that these hypothetical errors would not be in the code written by a user.

Reply:
I’m just thinking that in general it’s nicer to throw errors than to crash because they tend to be more actionable for the user. Either because they re-declared the macro or because there was some hole that we missed. But up to you.